Reindex share retry between hit sources

henningandersen opened the pull request:
The client and remote hit sources had each their own retry mechanism,
which would do the same. Supporting resiliency we would have to expand
on the retry mechanisms and as a preparation for that, the retry
mechanism is now shared such that each sub class is only responsible for
sending requests and converting responses/failures to common format.

Part of #42612

henningandersen committed:
Reindex share retry between hit sources

The client and remote hit sources had each their own retry mechanism,
which would do the same. Supporting resiliency we would have to expand
on the retry mechanisms and as a preparation for that, the retry
mechanism is now shared such that each sub class is only responsible for
sending requests and converting responses/failures to common format.

Part of #42612

elasticmachine commented:
Pinging @elastic/es-distributed

henningandersen committed:
Minor code improvement.

henningandersen commented:
@elasticmachine update-branch

henningandersen commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enhance_reindex_shared_retry

henningandersen commented:
Failure looks idential to #43889 but for another test,
@elasticmachine run elasticsearch-ci/2

tbrooks8 commented on server/src/main/java/org/elasticsearch/index/reindex/RejectAwareActionListener.java:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * Return a new listener that delegates failure/reject to original but forwards response to responseHandler
+     */
+    default <X> RejectAwareActionListener<X> withResponseHandler(Consumer<X> responseHandler) {
```
My preference is for `default` to be used in a scenario where you are providing default behavior for an interface that some implementation *might* want to overwrite. This feels more like a helper static method such as:

```java
static <X> RejectAwareActionListener<X> withResponseHandler(RejectAwareActionListener<X> errorDelegate, Consumer<X> responseHandler) {
// Implementation
}
```

henningandersen replied:
I agree. Fixed in 07d479303e7b6add9cdb5deea506dafeaa3cd7ed

Tim-Brooks reviewed (approved):
LGTM - I added a comment related to my preferences. But if you feel strongly otherwise, I'm fine with it.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into enhance_reindex_shared_retry

henningandersen committed:
Prefer static over default

No intention to override witResponsseHandler method so static is more
appropriate.

henningandersen committed:
Merge branch 'enhance_reindex_shared_retry' of github.com:henningandersen/elasticsearch into enhance_reindex_shared_retry

henningandersen committed:
Variable naming.
